artifact/sha/platform/openssl/sha.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <cstdint>
+#include <string>
+#include <ctime>
+#include <iomanip>
+#include <vector>
+#include <sstream>
+#include <algorithm>
+
+#include <openssl/evp.h>
+#include <artifact/sha/sha.hpp>
+
+#include <common/common.hpp>
+
+
+namespace mender {
+namespace sha {
+
+namespace log = mender::common::log;
+
+const ErrorCategoryClass ErrorCategory = ErrorCategoryClass();
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "ShaSumErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	InitializationError:
+		return "Initialization error";
+	ShasumCreationError:
+		return "Shasum creation error";
+	ShasumMismatchError:
+		return "Shasum mismatch error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+Reader::Reader(io::Reader &reader) :
+	Reader::Reader {reader, ""} {
+}
+
+Reader::Reader(io::Reader &reader, const std::string &expected_sha = "") :
+	sha_handle_(EVP_MD_CTX_new(), [](EVP_MD_CTX *ctx) { EVP_MD_CTX_free(ctx); }),
+	wrapped_reader_ {reader},
+	buffer_(4092),
+	expected_sha_ {expected_sha} {
+	if (EVP_DigestInit_ex(sha_handle_.get(), EVP_sha256(), nullptr) != 1) {
+		log::Error("Failed to initialize the shasummer");
+		initialized_ = false;
+		return;
+	}
+	initialized_ = true;
+}
+
+// Simply takes a reference to a buffer, and if need be resizes it, and then on
+// destruction, if it has been resized, sets it back to the original size
+class BufferResizer {
+	vector<uint8_t> &buf;
+	int original_size;
+
+public:
+	BufferResizer(vector<uint8_t> &buf, int original_size, int max_size) :
+		buf {buf},
+		original_size {original_size} {
+		buf.resize(std::min(original_size, max_size));
+	}
+
+	~BufferResizer() {
+		buf.resize(original_size);
+	}
+};
+
+expected::ExpectedSize Reader::Read(vector<uint8_t> &dst) {
+	if (!initialized_) {
+		return MakeError(InitializationError, "");
+	}
+
+	auto buffer_resizer = BufferResizer(buffer_, buffer_.size(), dst.size());
```

Comment:
Wouldn't it better to allow access to the buffer of the `wrapped_reader`? Why do we need another buffer here and copy things?

Reply:
Hmm, it's actually a really good point I think!

I can just copy into `dst` directly, and use it internally! 

Good :eye:

Reply:
Note: the BufferResizer is being removed in #1164.

Reply:
It is also removed in: https://github.com/mendersoftware/mender/pull/1166